GraphicsContextCG should clip the paths via CGContext in GPUP mode

kkinnunen-apple opened the pull request:
#### a8f57b13f2c7937bbd719183f5e60ff868433d87
<pre>
GraphicsContextCG should clip the paths via CGContext in GPUP mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=252833">https://bugs.webkit.org/show_bug.cgi?id=252833</a>
<a href="https://rdar.apple.com/105835901">rdar://105835901</a>

Reviewed by Said Abou-Hallawa.

Construct the clipped path directly into the CGContext.
Instead of creating a CGPath object, add the path info upon need to
the CGContext.

Removes some of CG work in GPUP receive side where time is spent
just copying the CGPath objects.

The implementation has now duplicated logic:
 - Add path segments to CGPath via PathCG
 - Add path segments to CGContext path

The implementation of both are stored next to each other so that
they keep in sync.

* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::setCGContextPath):
(WebCore::drawPathWithCGContext):
(WebCore::GraphicsContextCG::drawNativeImageInternal):
(WebCore::GraphicsContextCG::drawPattern):
(WebCore::GraphicsContextCG::drawPath):
(WebCore::GraphicsContextCG::fillPath):
(WebCore::GraphicsContextCG::strokePath):
(WebCore::GraphicsContextCG::fillRect):
(WebCore::GraphicsContextCG::fillRectWithRoundedHole):
(WebCore::GraphicsContextCG::clipOut):
(WebCore::GraphicsContextCG::clipPath):
(WebCore::GraphicsContextCG::beginTransparencyLayer):
(WebCore::GraphicsContextCG::drawLinesForText):
* Source/WebCore/platform/graphics/cg/PathCG.cpp:
(WebCore::addToContextPath):
(WebCore::copyClosingSubpathsApplierFunction):
(WebCore::addToCGContextPath):
* Source/WebCore/platform/graphics/cg/PathCG.h:

Canonical link: <a href="https://commits.webkit.org/271354@main">https://commits.webkit.org/271354@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/03811ed1aa159eeb815a498e76768dea39e73882

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27878 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6517 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29153 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30409 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25445 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/28372 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8493 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3911 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/25215 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28144 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5276 "Found 1 new test failure: editing/execCommand/insert-ordered-list-and-delete.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23931 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4546 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4735 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24946 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31097 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25482 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25389 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30935 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4742 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2897 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28798 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6263 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6738 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5188 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5210 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa reviewed (changes requested):
Can you please make this PR only for adding the path segments to `CGContext` path? In other words, add only what is needed to implement `addToCGContextPath()`.

The changes for adding the path segments to `CGPath` via `PathCG` can be moved to another PR.

shallawa commented on Source/WebCore/platform/graphics/cg/PathCG.cpp:
```diff
@@ -130,16 +164,38 @@ void PathCG::add(PathEllipse ellipse)
+    CGAffineTransform oldTransform = CGContextGetCTM(context);
+    AffineTransform transform;
+    transform.translate(segment.center.x(), segment.center.y()).rotate(rad2deg(segment.rotation)).scale(segment.radiusX, segment.radiusY);
+    CGAffineTransform cgTransform = transform;
```
Why do we need this? `AffineTransform` has the operator `CGAffineTransform()`.

shallawa commented on Source/WebCore/platform/graphics/cg/PathCG.cpp:
```diff
@@ -130,16 +164,38 @@ void PathCG::add(PathEllipse ellipse)
+{
+    CGAffineTransform oldTransform = CGContextGetCTM(context);
+    AffineTransform transform;
+    transform.translate(segment.center.x(), segment.center.y()).rotate(rad2deg(segment.rotation)).scale(segment.radiusX, segment.radiusY);
```
The same code is in `PathCG::add(PathEllipse)`. Can it be shared?

shallawa commented on Source/WebCore/platform/graphics/cg/PathCG.cpp:
```diff
@@ -445,6 +520,52 @@ FloatRect PathCG::strokeBoundingRect(const Function<void(GraphicsContext&)>& str
+        addToCGContextPath(context, *singleSegment);
+        return;
+    }
+    if (auto* segments = path.segmentsIfExists(); LIKELY(segments)) {
```
What is the purpose of this pattern `LIKELY(segments)`? And how can it be likely after checking `LIKELY(singleSegment)` above. How can this help branch prediction?

kkinnunen-apple replied:
This function is called for the purpose of it to use the segments, put the segments directly to the CGContext. This function would not be called if we didn't expect most cases to have the segments.

First likely:
It is more likely that the path has single segment than that it does not have.

Second likely, in case the path did not have a single segment:
It is more likely that the path has segments than that it does not have.

```
if (LIKELY(a))
    return "a";
if (LIKELY(b))
   return "b";
```
In this pattern, the "LIKELY" parts do not relate to each other. They relate to "is this branch likely to be taken or not".
Both can be likely.

shallawa commented on Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
```diff
@@ -167,6 +168,22 @@ static void setCGBlendMode(CGContextRef context, CompositeOperator op, BlendMode
+
+static void drawPathWithCGContext(CGContextRef context, CGPathDrawingMode drawingMode, const Path& path)
+{
+#if HAVE(CG_CONTEXT_DRAW_PATH_DIRECT)
+    CGContextDrawPathDirect(context, drawingMode, path.platformPath(), nullptr);
+#else
+    setCGContextPath(context, path);
+    CGContextDrawPath(context, drawingMode);
+#endif
```
I think we should remove `HAVE(CG_CONTEXT_DRAW_PATH_DIRECT)` and the `#else` part since HAVE_CG_CONTEXT_DRAW_PATH_DIRECT is true for all `PLATFORM(COCOA)`. Maybe in another patch is fine. But this means drawPath() does not benefit for this change.

kkinnunen-apple replied:
> But this means drawPath() does not benefit for this change.

Yeah, this change is not a progression with drawPath for our current benchmarks.

kkinnunen-apple replied:
To add, as much as I could gather, this is due to when we add segments via CGContextAdd*, the segments are transformed with CGContext CTM. This seems to be slow when done per call.

When creating an untransformed path and then drawing it directly, the draw call can do the transform book keeping once and then iterate over the path segments, transforming as it goes. Perhaps also some paths don't need the transform, rather the transform can be an argument.

shallawa commented on Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
```diff
@@ -167,6 +168,22 @@ static void setCGBlendMode(CGContextRef context, CompositeOperator op, BlendMode
+    addToCGContextPath(context, path);
+}
+
+static void drawPathWithCGContext(CGContextRef context, CGPathDrawingMode drawingMode, const Path& path)
```
I think adding `WithCGContext` is redundant and uncommon since it can be deduced from the first argument.

kkinnunen-apple replied:
The name resolution doesn't work this way. GraphicsContextCG::drawPath conflicts with this.
I'll land with this name.
I'll remove the function altogether when removing the ifdef.

shallawa commented on Source/WebCore/platform/graphics/cg/PathCG.cpp:
```diff
@@ -206,11 +262,30 @@ void PathCG::add(PathRoundedRect roundedRect)
+    path->add(WTFMove(segment));
+    // CGContextAddPath has a bug with existing MoveToPoints in context path.
+    // rdar://118395262
+    auto ctm = CGContextGetCTM(context);
+    auto transformedPath = adoptCF(CGPathCreateCopyByTransformingPath(path->platformPath(), &ctm));
+    CGContextSetCTM(context, CGAffineTransformIdentity);
+    CGContextAddPath(context, transformedPath.get());
+    CGContextSetCTM(context, ctm);
```
So this is a workaround for rdar://118395262?

kkinnunen-apple replied:
Yes, the transform part.

shallawa commented on Source/WebCore/platform/graphics/cg/PathCG.cpp:
```diff
@@ -130,16 +164,38 @@ void PathCG::add(PathEllipse ellipse)
+    CGContextConcatCTM(context, cgTransform);
+    // CG coordinates system increases the angle in the anticlockwise direction.
+    CGContextAddArc(context, 0, 0, 1, segment.startAngle, segment.endAngle, segment.direction == RotationDirection::Counterclockwise);
+    CGContextSetCTM(context, oldTransform);
```
Should we use similar technique like what you did for the rounded rect? I think using PathEllipse (not PathEllipseInRect) is not very common.

```
auto path = PathCG::create();
path->add(WTFMove(segment));
```

kkinnunen-apple replied:
I think with the reused transform code, maybe it's not that complicated to have it duplicated?

kkinnunen-apple committed:
GraphicsContextCG should clip the paths via CGContext in GPUP mode
https://bugs.webkit.org/show_bug.cgi?id=252833
rdar://105835901

Reviewed by Said Abou-Hallawa.

Construct the clipped path directly into the CGContext.
Instead of creating a CGPath object, add the path info upon need to
the CGContext.

Removes some of CG work in GPUP receive side where time is spent
just copying the CGPath objects.

The implementation has now duplicated logic:
 - Add path segments to CGPath via PathCG
 - Add path segments to CGContext path

The implementation of both are stored next to each other so that
they keep in sync.

* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::setCGContextPath):
(WebCore::drawPathWithCGContext):
(WebCore::GraphicsContextCG::drawNativeImageInternal):
(WebCore::GraphicsContextCG::drawPattern):
(WebCore::GraphicsContextCG::drawPath):
(WebCore::GraphicsContextCG::fillPath):
(WebCore::GraphicsContextCG::strokePath):
(WebCore::GraphicsContextCG::fillRect):
(WebCore::GraphicsContextCG::fillRectWithRoundedHole):
(WebCore::GraphicsContextCG::clipOut):
(WebCore::GraphicsContextCG::clipPath):
(WebCore::GraphicsContextCG::beginTransparencyLayer):
(WebCore::GraphicsContextCG::drawLinesForText):
* Source/WebCore/platform/graphics/cg/PathCG.cpp:
(WebCore::addToContextPath):
(WebCore::copyClosingSubpathsApplierFunction):
(WebCore::addToCGContextPath):
* Source/WebCore/platform/graphics/cg/PathCG.h:

Canonical link: https://commits.webkit.org/271354@main
